Rent vs mortgage maps (#18)
yimbymelbourne/data: src/rent-vs-mortgage.md

</div>

> ## Where are the mortages relative "affordable" (closest to rents)?

Reviewer: ```suggestion
## Where are the mortgages relatively "affordable" (closest to rents)?
```

```js

> const propertyType = view(Inputs.radio(['HOUSE', 'UNIT'], { label: 'Property Type', value: 'HOUSE' }))
> const rangeType = view(Inputs.radio(['DYNAMIC', 'FIXED'], { label: 'Colour Range Type', value: 'FIXED' }))

Reviewer: ```suggestion
const propertyType = view(Inputs.radio(['HOUSE', 'UNIT'], { label: 'Property type', value: 'HOUSE' }))
const rangeType = view(Inputs.radio(['DYNAMIC', 'FIXED'], { label: 'Colour range type', value: 'FIXED' }))
```

```

> The FIXED colour range option ensures the same fixed range of colours for consistent display across both units and houses, while the DYNAMIC option adjusts the range based on the data to maximise the contrast.

Reviewer: ```suggestion
The `FIXED` colour range option ensures the same fixed range of colours for consistent display across both units and houses, while the `DYNAMIC` option adjusts the range based on the data to maximise the contrast.
```

```js
const dataArray = data.toArray().filter(row => row.property_type === propertyType)
const sa2Map = new Map(dataArray.map(row => [row.sa2_code.toString(), row]))
const sa2Codes = new Set(sa2Map.keys())
const domain = rangeType === 'FIXED'
   ? FIXED_DOMAIN
   : dataArray.reduce((acc, row) => {
        const ratio = row.repayment_to_rent_ratio
        return [Math.min(acc[0], ratio), Math.max(acc[1], ratio)]
      },
      [100, 0],

>    )

Reviewer: shouldn't Plot dynamically compute the domain for the color scale here? We could yield `undefined` on the else condition if that's the case

Author: Oh, that's a great point - definitely easier to leave it to Plot in that case, I'll check it out.

Author: Yeah that works, just need to actually plug in some data into the unrendered plot below that I use to get the range.

> const deckColours = Plot.plot({ color: { domain } }).scale('color')
> const plotRatioLegend = Plot.legend({
>   label: 'Repayment to rent ratio',
>   color: deckColours,
> })
> ```

Reviewer: ```suggestion
const deckColors = Plot.plot({ color: { domain } }).scale('color')
const plotRatioLegend = Plot.legend({
  label: 'Repayment to rent ratio',
  color: deckColors,
})
```
```

Author: I refuse to give in to American cultural hegemony :D

  ],
  controller: true,
  map: mapboxgl,
  mapboxApiAccessToken: 'pk.eyJ1IjoibWF4Ym8iLCJhIjoiY2lsd2JnZDRyMDFxNHZna3MwZDZmN2R0ZCJ9.32XWATCazaiDzdW6bXvBxw',
  mapStyle: 'mapbox://styles/mapbox/light-v10',
  layers: [
    new GeoJsonLayer({
      id: 'sa2s',
      data: melbourneSa2s,
      pickable: true,
      getFillColor: (f) => {
        const sa2 = sa2Map.get(f?.properties?.REGION_CODE)
        if (!sa2) return

        const hex = deckColours.apply(sa2.repayment_to_rent_ratio)
        const color = Color(hex)

        return [color.red * 255, color.green * 255, color.blue * 255, 0.6 * 255]
      },
      getLineColor: 0,
      getLineWidth: 20,
    }),
  ],

>   getTooltip: (point) => {

Reviewer: ah awesome, I wasn't quite sure why this wasn't working

    const properties = point?.object?.properties
    if (!properties) return

    const sa2 = sa2Map.get(properties.REGION_CODE)
    return sa2
      ? `${properties.REGION_NAME}\nMortgage ${percentFormat(sa2.repayment_to_rent_ratio)} of Rent`
      : properties.REGION_NAME
  },
})

// clean up if this code re-runs
invalidation.then(() => {
  deckInstance.finalize()
  container.innerHTML = ''
})
```

<div class='card'>

```js
Plot.plot({
  aspectRatio: 1,
  color: {
    legend: true,
    domain,
    label: 'Repayment to rent ratio',
  },
  marks: [
    Plot.geo(melbourneSa2s, {

>       fill: f => sa2Map.get(f?.properties?.REGION_CODE)?.repayment_to_rent_ratio,

Reviewer: we might want to add a `stroke: "black"` to this if want the Deck and Plot maps to look the same

Author: Yeah, better like this for my tastes.

![image](https://github.com/user-attachments/assets/02a6ae15-e7cf-49cd-ac0c-1a25d2ff2613)
